File: frontend/lib/phases.ts
/**
 * Progress phases — mirrors `orchestrator/app/core/sf_intel/phases.py`.
 *
 * The indicator is driven ENTIRELY by what the backend emits. There is no
 * timer walking through plausible-looking steps, and no phase is ever inferred:
 * if the server has not said it is querying Salesforce, the UI does not say so
 * either. That is the difference between a progress indicator and a decoration.
 */

export const PHASES = [
  'understanding',
  'resolving_context',
  'checking_schema',
  'clarifying',
  'querying_salesforce',
  'retrieving_more_results',
  'analyzing_records',
  'calculating',
  'verifying',
  'drafting_answer',
  'reconnecting',
  'completed',
  'failed',
] as const;

export type Phase = (typeof PHASES)[number];

/** Live status carried on the `status` SSE event and the final `meta`. */
export interface PhaseStatus {
  phase: Phase;
  label: string;
  run_id?: string;
  started_at?: string;
  record_count?: number;
  tool_name?: string;
}

const PHASE_SET = new Set<string>(PHASES);

/** Phases during which the star animates. Terminal ones stop it. */
const ACTIVE = new Set<Phase>(
  PHASES.filter((p) => p !== 'completed' && p !== 'failed' && p !== 'clarifying'),
);

export function isPhase(value: unknown): value is Phase {
  return typeof value === 'string' && PHASE_SET.has(value);
}

export function isActivePhase(phase: Phase | undefined): boolean {
  return phase !== undefined && ACTIVE.has(phase);
}

/**
 * The visual family a phase belongs to. Six states, not thirteen: the star
 * changes character when the KIND of work changes, and a user watching
 * "Analyzing 42 records" become "Verifying the totals" should not see the
 * animation restart.
 */
export type StarState =
  | 'understanding'
  | 'searching'
  | 'calculating'
  | 'verifying'
  | 'drafting'
  | 'reconnecting';

const STAR_STATE: Record<Phase, StarState | null> = {
  understanding: 'understanding',
  resolving_context: 'understanding',
  checking_schema: 'searching',
  clarifying: null,
  querying_salesforce: 'searching',
  retrieving_more_results: 'searching',
  analyzing_records: 'calculating',
  calculating: 'calculating',
  verifying: 'verifying',
  drafting_answer: 'drafting',
  reconnecting: 'reconnecting',
  completed: null,
  failed: null,
};

export function starState(phase: Phase | undefined): StarState | null {
  return phase ? STAR_STATE[phase] : null;
}

/**
 * Screen-reader text. Announced through aria-live, so it says what is HAPPENING
 * rather than restating the visible label word for word.
 */
export function accessibleStatus(status: PhaseStatus | null): string {
  if (!status) return '';
  if (status.phase === 'completed') return 'Answer ready.';
  if (status.phase === 'failed') return 'The request did not complete.';
  return `Salesforce assistant is processing the request: ${status.label}`;
}

/**
 * Parse a `status` SSE payload into a typed phase, or null.
 *
 * `text` alone (no `phase`) is the pre-existing web-search/URL progress line —
 * it stays a plain string and does NOT drive the star, so the two progress
 * systems cannot fight over the same row.
 */
export function parsePhaseStatus(raw: unknown): PhaseStatus | null {
  if (!raw || typeof raw !== 'object') return null;
  const s = raw as Record<string, unknown>;
  if (!isPhase(s.phase)) return null;
  return {
    phase: s.phase,
    label: typeof s.text === 'string' ? s.text : s.phase,
    ...(typeof s.run_id === 'string' ? { run_id: s.run_id } : {}),
    ...(typeof s.started_at === 'string' ? { started_at: s.started_at } : {}),
    ...(typeof s.record_count === 'number'
      ? { record_count: s.record_count }
      : {}),
    ...(typeof s.tool_name === 'string' ? { tool_name: s.tool_name } : {}),
  };
}

File: orchestrator/app/core/sf_intel/phases.py
"""Progress phases — what the UI is allowed to say while a request runs.

These are SUMMARIES of work that has actually started, not a narration of the
model's thinking and not a timer walking through plausible-sounding steps. Each
one is emitted at the moment the corresponding stage begins, so "Searching
Salesforce" means a query is in flight, and the animation stops when the
backend stops.

They ride on the EXISTING `status` SSE event, additively: the payload keeps its
`{"text": ...}` shape (every current client reads that and nothing else) and
gains `phase`, `run_id` and the optional counters. No new event name, so replay,
persistence and the frontend's allowlist are untouched.
"""
from __future__ import annotations

from typing import Optional

#: Ordered roughly as a request meets them. `reconnecting` and the two terminal
#: phases are out-of-band and may follow any of the others.
PHASES = (
    "understanding",
    "resolving_context",
    "checking_schema",
    "clarifying",
    "querying_salesforce",
    "retrieving_more_results",
    "analyzing_records",
    "calculating",
    "verifying",
    "drafting_answer",
    "reconnecting",
    "completed",
    "failed",
)

#: The default user-facing wording. Concise and factual — a label that promises
#: work the backend is not doing is worse than no label.
LABELS = {
    "understanding": "Understanding your request",
    "resolving_context": "Using this conversation's context",
    "checking_schema": "Checking Salesforce fields",
    "clarifying": "Checking one detail with you",
    "querying_salesforce": "Searching Salesforce",
    "retrieving_more_results": "Retrieving more results",
    "analyzing_records": "Analyzing records",
    "calculating": "Calculating the totals",
    "verifying": "Verifying the numbers",
    "drafting_answer": "Preparing the answer",
    "reconnecting": "Reconnecting",
    "completed": "Done",
    "failed": "That did not work",
}

#: Phases during which the UI shows an active indicator. `completed`/`failed`
#: are terminal, and `clarifying` hands over to the card rather than spinning.
ACTIVE_PHASES = frozenset(PHASES) - {"completed", "failed", "clarifying"}


def label_for(phase: str, *, record_count: Optional[int] = None) -> str:
    """The line the user reads. Counts are folded in where they are real.

    "Analyzing 42 records" is only allowed to say 42 when 42 records exist —
    the count comes from the query result, never from an estimate.
    """
    base = LABELS.get(phase, "Working")
    if record_count is None:
        return base
    if phase == "analyzing_records":
        return f"Analyzing {record_count:,} record{'' if record_count == 1 else 's'}"
    if phase == "retrieving_more_results":
        return f"Retrieved {record_count:,} record{'' if record_count == 1 else 's'} so far"
    return base


def status_payload(
    phase: str,
    *,
    run_id: str = "",
    started_at: str = "",
    record_count: Optional[int] = None,
    tool_name: str = "",
    label: str = "",
) -> dict:
    """The `status` event payload for one phase.

    `text` stays first-class and is what pre-existing clients render; the rest
    is additive metadata a newer client uses to drive the phase indicator.
    """
    if phase not in PHASES:
        raise ValueError(f"unknown phase {phase!r} (allowed: {', '.join(PHASES)})")
    payload = {
        "text": label or label_for(phase, record_count=record_count),
        "phase": phase,
    }
    if run_id:
        payload["run_id"] = run_id
    if started_at:
        payload["started_at"] = started_at
    if record_count is not None:
        payload["record_count"] = int(record_count)
    if tool_name:
        payload["tool_name"] = tool_name
    return payload
